tutorials/haskell/golden-testing-in-haskell-with-hspec-gold/tutorial.md

```diff
@@ -0,0 +1,190 @@
+This code sample will help us to understand how does basic golden testing works.
+
+##Installing the library:
+To install the [hspec-golden](https://github.com/stackbuilders/hspec-golden) superset we need to run this command on the console:
+
+``$ stack install hspec-golden``
+
+##Importing the libraries
+In our project on we have created a GoldenSpec.hs and added the following lines.
+``` 
+module GoldenSpec where
+
+import Test.Hspec
+import Test.Hspec.Golden
+import qualified Data.ByteString as B
+import qualified Data.ByteString.Lazy as BL
+import Data.Yaml as Y
+import Data.Aeson as A
+import Control.Exception (evaluate)
+```
+
+Here we are telling Haskell to add the libraries that are going to be implemented in this tutorial,
+we are adding the test framework [hspec](https://github.com/hspec/hspec) and [hspec-golden](https://github.com/stackbuilders/hspec-golden),
+with some additional libs to implement our own Golden type.
+
+
+###Writing our first golden test
+The [hspec-golden](https://github.com/stackbuilders/hspec-golden) library provides us with an easy example to test the functionality of the library with the `defaultGolden` helper:
+
+``` 
+describe "id" $
+     it "generates the right output with the right params" $
+        let output = show $ id (1:: Int)
+        in defaultGolden "id" output
+```
+
+On this test we are setting up a B.D.D. test with [hspec](https://github.com/hspec/hspec) “describe” and “it”,
+but on the inside there is an implementation of the `defaultGolden` helper,
+that will create(or update) for us a directory called `.golden` this will contain another directory named after our described test in this case `id` with two files containing the `actual` and `golden` values for our test.
+
+For our example we will create our own Golden type.
+We have created a file called test.yml that contains an ansible role on the dir ["test/fixtures/"](https://github.com/MedAdrian/hspec-golden-tutorial/blob/master/test/fixtures/test.yml) 
+we will use this file contents and transform them into JSON.
+
+```
+{--
+Here we define our Golden test as follows:
+- We read a .yml file that contains an ansible role
+- We decode the string read by Haskell that we read form the .yml file
+--}
+
+spec :: Spec
+spec = do
+    describe "conversion" $ do
+      content <- runIO $ B.readFile "test/fixtures/test.yml"
+      let eYaml = Y.decodeEither' content :: Either Y.ParseException Y.Value  --Parses to a generic yml type.
+          Right json = A.encode <$> eYaml                                     --Parses the data type to a json structure.
+      it "transforms a YAML file contents to JSON" $ do
+        myGoldenTest "conversion" json                                        --Generates the golden test.
+```
+
+To see the full code example access this [rerpository](https://github.com/MedAdrian/hspec-golden-tutorial).
+
+If you are confused about how test are written you can follow this [link](http://hspec.github.io/hspec-discover.html).
```

Comment:
@MedAdrian Probably, this is out of the scope of this tutorial. Is it worth mention it?

Reply:
I thought it is, since we may have some people new to haskell as i am, i did not know what `{-# OPTIONS_GHC -F -pgmF hspec-discover #-}`do on the Spec.hs file, but please tell me if you think we should remove this.

Reply:
I think we should consider on removing it. I think mention that is more related to `hspec` that just `hspec-golden`. Well, I'll leave it up to you. :slightly_smiling_face:

Reply:
I've removed it.